lib/modules/firestore/TransactionHandler.js

```diff
@@ -0,0 +1,231 @@
+    };
+
+    meta.transaction = new Transaction(this._firestore, meta);
+    this._pending[id] = meta;
+
+    // deferred promise
+    return new Promise((resolve, reject) => {
+      getNativeModule(this._firestore).transactionBegin(id);
+      meta.resolve = r => {
+        resolve(r);
+        this._remove(id);
+      };
+      meta.reject = e => {
+        reject(e);
+        this._remove(id);
+      };
+    });
+  }
+
+  /**
+   * Destroys a local instance of a transaction meta
+   *
+   * @param id
+   * @param pendingAbort Notify native that there's still an transaction in
+   *        progress that needs aborting - this is to handle a JS side
+   *        exception
+   * @private
+   */
+  _remove(id, pendingAbort = false) {
+    // todo confirm pending arg no longer needed
+    getNativeModule(this._firestore).transactionDispose(id, pendingAbort);
+    // TODO may need delaying to next event loop
+    delete this._pending[id];
+  }
+
+  /**
+   * -------------
+   *    EVENTS
+   * -------------
+   */
+
+  /**
+   * Handles incoming native transaction events and distributes to correct
+   * internal handler by event.type
+   *
+   * @param event
+   * @returns {*}
+   * @private
+   */
+  _handleTransactionEvent(event: TransactionEvent) {
+    switch (event.type) {
+      case 'update':
+        return this._handleUpdate(event);
+      case 'error':
+        return this._handleError(event);
+      case 'complete':
+        return this._handleComplete(event);
+      default:
+        getLogger(this._firestore).warn(
+          `Unknown transaction event type: '${event.type}'`,
+          event
+        );
+        return undefined;
+    }
+  }
+
+  /**
+   * Handles incoming native transaction update events
+   *
+   * @param event
+   * @private
+   */
+  async _handleUpdate(event: TransactionEvent) {
+    const { id } = event;
+    // abort if no longer exists js side
+    if (!this._pending[id]) return this._remove(id);
+
+    const { updateFunction, transaction, reject } = this._pending[id];
+
+    // clear any saved state from previous transaction runs
+    transaction._prepare();
+
+    let finalError;
+    let updateFailed;
+    let pendingResult;
+
+    // run the users custom update functionality
+    try {
+      const possiblePromise = updateFunction(transaction);
+
+      // validate user has returned a promise in their update function
+      // TODO must it actually return a promise? Can't find any usages of it without one...
+      if (!possiblePromise || !possiblePromise.then) {
+        finalError = new Error(
+          'Update function for `firestore.runTransaction(updateFunction)` must return a Promise.'
+        );
+      } else {
+        pendingResult = await possiblePromise;
+      }
+    } catch (exception) {
+      updateFailed = true; // in case the user rejects with nothing
+      finalError = exception;
+    }
+
+    // reject the final promise and remove from native
+    if (updateFailed) {
```

Comment:
At the moment, this only exits if the promise throws an exception, not if the promise isn't actually a promise. Is this correct?

Reply:
Exits when the update function throws an error (e.g. bad code) or it Promise rejects. Added this a a boolean flag instead of `if (finalError) { } ` as it's possible to `Promise.reject()` with no value/exception which would fall through and not exit.

Reply:
Ok, was just checking I could only see the boolean being set to true in the try/catch